tt_eager/tt_dnn/op_library/rotary_embedding/rotary_embedding_llama_op.hpp

```diff
@@ -0,0 +1,95 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <functional>
+
+#include "tensor/tensor.hpp"
+#include "third_party/magic_enum/magic_enum.hpp"
+#include "tt_dnn/op_library/run_operation.hpp"
+#include "tt_metal/host_api.hpp"
+#include "tt_dnn/op_library/compute_kernel_config.hpp"
+
+namespace tt {
+
+namespace tt_metal {
+
+enum class RotaryEmbeddingLlamaOpParallelizationStrategy { MULTI_CORE = 0, SINGLE_CORE = 1 };
+
+operation::ProgramWithCallbacks rotary_embedding_llama_single_core(
+    const Tensor &input, const Tensor &cos, const Tensor &sin, const Tensor &trans_mat, Tensor &output, DeviceComputeKernelConfig compute_kernel_config);
+operation::ProgramWithCallbacks rotary_embedding_llama_multi_core(
+    const Tensor &input, const Tensor &cos, const Tensor &sin, const Tensor &trans_mat, Tensor &output, DeviceComputeKernelConfig compute_kernel_config);
+
+struct RotaryEmbeddingLlama {
+    const uint32_t seq_len;
+    const MemoryConfig output_mem_config;
+    const DeviceComputeKernelConfig compute_kernel_config;
+
+    RotaryEmbeddingLlamaOpParallelizationStrategy get_parallelization_strategy(
+        const std::vector<Tensor> &input_tensors) const;
+
+    void validate(const std::vector<Tensor> &input_tensors) const;
+    std::vector<Shape> compute_output_shapes(const std::vector<Tensor> &input_tensors) const;
+    std::vector<Tensor> create_output_tensors(const std::vector<Tensor> &input_tensors) const;
+
+    operation::ProgramWithCallbacks create_program(
+        const std::vector<Tensor> &input_tensors, std::vector<Tensor> &output_tensors) const;
+    tt::stl::reflection::Attributes attributes() const;
+
+    const operation::Hash compute_program_hash(const std::vector<Tensor> &input_tensors) const;
+};
+
+inline Tensor rotary_embedding_llama(
+    const Tensor &input_tensor,
+    const Tensor &cos,
+    const Tensor &sin,
+    const Tensor trans_mat,
+    const MemoryConfig &output_mem_config = operation::DEFAULT_OUTPUT_MEMORY_CONFIG,
+    std::optional<const DeviceComputeKernelConfig> compute_kernel_config = std::nullopt) {
+    std::vector<Tensor> output_tensors = {Tensor(operation::get_workers_for_op_output({input_tensor, cos, sin, trans_mat}))};
+    operation::launch_with_autoformat(
+        [output_mem_config, compute_kernel_config] (const std::vector<Tensor>& input_tensors, const std::vector<std::optional<const Tensor>>& optional_input_tensors, const std::vector<std::optional<Tensor>>& optional_output_tensors) mutable -> std::vector<Tensor> {
+            auto& input_tensor = input_tensors.at(0);
+            auto& cos = input_tensors.at(1);
+            auto& sin = input_tensors.at(2);
+            auto& trans_mat = input_tensors.at(3);
+            TT_FATAL(input_tensor.get_legacy_shape()[-1] % TILE_WIDTH == 0, "Input X dim must be divisible into tiles");
```

Comment:
There are too many checks here. Please use the validate function of the op to perform these checks. 
This will become an e2e performance issue since this code is on the critical path in the worker threads.